### ISSUE TITLE
EKSA Bare Metal: cluster ScaleUp operation fails because of provided cert bundles (under hostOsConfiguration of tinkerbellMachineConfig kind)

### ISSUE THREAD

**Issue by hiteshpattanayak-rafay** [aws/eks-anywhere#6108]
**what happened:**
Scale Up operation failed for a 2 node EKSA Bare Metal cluster (1 control-plane node, 1 worker node : both running bottlerocket-os) where the Kubelet failed to start on newly added bottlerocket-os worker node.

It is to be noted that the machine config of both control plane and worker node groups have cert bundles under host os configuration.
The scale operation fails when we provide above host os configuration. Else it works just fine.

**eksctl anywhere version:** v0.16.1

steps to reproduce:
1. Provision a cluster (name used for this: jul1cluster) with 1 cp and 1 dp node.
2. Create a spare hardware with label selector matching with existing worker node group.
3. Scale up worker node group using command `eksctl/eksctl anywhere upgrade cluster -f eksa-jul1cluster-cluster-scale.yaml --force-cleanup`
note: hardware(s) not specified as spare hardware CRDs already exists.

**please find cli logs for the same**
```
Warning: The recommended number of control plane nodes is 3 or 5
Warning: The recommended number of control plane nodes is 3 or 5
Warning: The recommended number of control plane nodes is 3 or 5
Performing setup and validations
✅ Tinkerbell provider validation
✅ Validate OS is compatible with registry mirror configuration
✅ Validate certificate for registry mirror
✅ Control plane ready
✅ Worker nodes ready
✅ Nodes ready
✅ Cluster CRDs ready
✅ Cluster object present on workload cluster
✅ Upgrade cluster kubernetes version increment
✅ Validate authentication for git provider
✅ Validate immutable fields
✅ Validate pod disruption budgets
Ensuring etcd CAPI providers exist on management cluster before upgrade
Pausing EKS-A cluster controller reconcile
Pausing GitOps cluster resources reconcile
Upgrading core components
Creating bootstrap cluster
Provider specific pre-capi-install-setup on bootstrap cluster
Provider specific post-setup
Installing cluster-api providers on bootstrap cluster
Backing up workload cluster's management resources before moving to bootstrap cluster
Moving management cluster from workload to bootstrap cluster
Upgrading workload cluster
collecting cluster diagnostics
collecting management cluster diagnostics
⏳ Collecting support bundle from cluster, this can take a while	{"cluster": "bootstrap-cluster", "bundle": "jul1cluster/generated/bootstrap-cluster-2023-07-01T13:17:41Z-bundle.yaml", "since": 1688206661950377815, "kubeconfig": "jul1cluster/generated/jul1cluster.kind.kubeconfig"}
Support bundle archive created	{"path": "support-bundle-2023-07-01T13_17_43.tar.gz"}
Analyzing support bundle	{"bundle": "jul1cluster/generated/bootstrap-cluster-2023-07-01T13:17:41Z-bundle.yaml", "archive": "support-bundle-2023-07-01T13_17_43.tar.gz"}
Analysis output generated	{"path": "jul1cluster/generated/bootstrap-cluster-2023-07-01T13:17:56Z-analysis.yaml"}
collecting workload cluster diagnostics
⏳ Collecting support bundle from cluster, this can take a while	{"cluster": "jul1cluster", "bundle": "jul1cluster/generated/jul1cluster-2023-07-01T13:18:02Z-bundle.yaml", "since": 1688206682474145329, "kubeconfig": "jul1cluster/jul1cluster-eks-a-cluster.kubeconfig"}
Support bundle archive created	{"path": "support-bundle-2023-07-01T13_18_04.tar.gz"}
Analyzing support bundle	{"bundle": "jul1cluster/generated/jul1cluster-2023-07-01T13:18:02Z-bundle.yaml", "archive": "support-bundle-2023-07-01T13_18_04.tar.gz"}
Analysis output generated	{"path": "jul1cluster/generated/jul1cluster-2023-07-01T13:18:26Z-analysis.yaml"}
Error: failed to upgrade cluster: applying capi control plane spec: executing apply: The KubeadmControlPlane "jul1cluster" is invalid: 
* spec.machineTemplate.infrastructure.apiVersion: Invalid value: "": cannot be empty
* spec.machineTemplate.infrastructure.kind: Invalid value: "": cannot be empty
* spec.machineTemplate.infrastructure.name: Invalid value: "": cannot be empty
* spec.version: Invalid value: "v": must be a valid semantic version
* spec.kubeadmConfigSpec.clusterConfiguration.certBundles: Forbidden: cannot be modified
* spec.kubeadmConfigSpec.clusterConfiguration.certificatesDir: Forbidden: cannot be modified
* spec.kubeadmConfigSpec.initConfiguration: Forbidden: cannot be modified
* spec.kubeadmConfigSpec.joinConfiguration: Forbidden: cannot be modified
* spec.version: Internal error: failed to parse updated kubeadmcontrolplane version: v: failed to parse major.minor.patch from "v"
```

**config used for scale operation:**
```
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: Cluster
metadata:
  name: jul1cluster
spec:
  kubernetesVersion: "1.26"
  clusterNetwork:
    cniConfig:
      cilium:
        policyEnforcementMode: default
    pods:
      cidrBlocks:
        - 192.168.0.0/16
    services:
      cidrBlocks:
        - 10.96.0.0/12
  controlPlaneConfiguration:
    count: 1
    endpoint:
      host: 192.168.10.197
    machineGroupRef:
      kind: TinkerbellMachineConfig
      name: mc1
  workerNodeGroupConfigurations:
    - count: 2
      name: md-0
      machineGroupRef:
        kind: TinkerbellMachineConfig
        name: mc2
  managementCluster:
    name: jul1cluster
  datacenterRef:
    kind: TinkerbellDatacenterConfig
    name: jul1cluster
---
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: TinkerbellDatacenterConfig
metadata:
  name: jul1cluster
spec:
  tinkerbellIP: 192.168.10.35
---
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: TinkerbellMachineConfig
metadata:
  name: mc1
spec:
  hardwareSelector:
    type: cp
  osFamily: bottlerocket
  templateRef:
    kind: TinkerbellTemplateConfig
  users:
    - name: ec2-user
      sshAuthorizedKeys:
        - ssh-rsa ..
  hostOSConfiguration:
    certBundles:
      - name: bundle1
        data: |-
          -----BEGIN CERTIFICATE-----
          ...
          -----END CERTIFICATE-----
---
apiVersion: anywhere.eks.amazonaws.com/v1alpha1
kind: TinkerbellMachineConfig
metadata:
  name: mc2
spec:
  hardwareSelector:
    type: dp
  osFamily: bottlerocket
  templateRef:
    kind: TinkerbellTemplateConfig
  users:
    - name: ec2-user
      sshAuthorizedKeys:
        - ssh-rsa ...
  hostOSConfiguration:
    certBundles:
      - name: bundle1
        data: |-
          -----BEGIN CERTIFICATE-----
          ...
          -----END CERTIFICATE-----
---

```

**available & occupied hardware** :
```
KUBECONFIG=jul1cluster/generated/jul1cluster.kind.kubeconfig kubectl get hardware -A --show-labels
NAMESPACE     NAME               STATE   LABELS
eksa-system   eksabmhp1-cp-n-1           type=cp,v1alpha1.tinkerbell.org/ownerName=jul1cluster-control-plane-template-1688208086793-2zj4t,v1alpha1.tinkerbell.org/ownerNamespace=eksa-system
eksa-system   eksabmhp1-dp-n-1           type=dp,v1alpha1.tinkerbell.org/ownerName=jul1cluster-md-0-1688208086794-tz2tp,v1alpha1.tinkerbell.org/ownerNamespace=eksa-system
eksa-system   eksabmhp1-dp-n-2           type=dp
```
**provisioned machines**:
```
KUBECONFIG=jul1cluster/generated/jul1cluster.kind.kubeconfig kubectl get machines.cluster.x-k8s.io -A
NAMESPACE     NAME                                      CLUSTER       NODENAME                                  PROVIDERID                                  PHASE     AGE   VERSION
eksa-system   jul1cluster-md-0-5fdb7c9d58x98kb9-sqppr   jul1cluster   jul1cluster-md-0-5fdb7c9d58x98kb9-sqppr   tinkerbell://eksa-system/eksabmhp1-dp-n-1   Running   23m   v1.26.4-eks-1-26-10
eksa-system   jul1cluster-t9gkj                         jul1cluster   jul1cluster-t9gkj                         tinkerbell://eksa-system/eksabmhp1-cp-n-1   Running   23m   v1.26.4-eks-1-26-10
```

Attached complete cli log: [eksa-cli.log](https://github.com/aws/eks-anywhere/files/11926409/eksa-cli.log)

**Comment by robbierafay** [comment]
- I can confirm I can reproduce the same issue on a two node (1 cp and 1 worker) self managed cluster when using hostOSConfiguration - certBundles in cluster config. 
- Impact is that cluster is left in an unusable state as no scale operation fails. 
- No Tinkerbell workflow gets created for the new node. 
- eksctl anywhere version v0.16.1

Noticed these errors in eksa-cli-logs file :
[2023-07-02T01_35_49.log.txt](https://github.com/aws/eks-anywhere/files/11928525/2023-07-02T01_35_49.log.txt)


`{"T":1688261913915760535,"M":"Executing command","cmd":"/usr/bin/docker exec -i eksa_1688261749647173861 kubectl apply -f - --namespace eksa-system --kubeconfig eksacert/generated/eksacert.kind.kubeconfig"}

{"T":1688261915272097645,"M":"docker","stderr":"The KubeadmControlPlane \"eksacert\" is invalid: \n* spec.machineTemplate.infrastructure.apiVersion: Invalid value: \"\": cannot be empty\n* spec.machineTemplate.infrastructure.kind: Invalid value: \"\": cannot be empty\n* spec.machineTemplate.infrastructure.name: Invalid value: \"\": cannot be empty\n* spec.version: Invalid value: \"v\": must be a valid semantic version\n* spec.kubeadmConfigSpec.clusterConfiguration.certBundles: Forbidden: cannot be modified\n* spec.kubeadmConfigSpec.clusterConfiguration.certificatesDir: Forbidden: cannot be modified\n* spec.kubeadmConfigSpec.initConfiguration: Forbidden: cannot be modified\n* spec.kubeadmConfigSpec.joinConfiguration: Forbidden: cannot be modified\n* spec.version: Internal error: failed to parse updated kubeadmcontrolplane version: v: failed to parse major.minor.patch from \"v\"\n"}

{"T":1688261915272131917,"M":"Error happened during retry","error":"executing apply: The KubeadmControlPlane \"eksacert\" is invalid: \n* spec.machineTemplate.infrastructure.apiVersion: Invalid value: \"\": cannot be empty\n* spec.machineTemplate.infrastructure.kind: Invalid value: \"\": cannot be empty\n* spec.machineTemplate.infrastructure.name: Invalid value: \"\": cannot be empty\n* spec.version: Invalid value: \"v\": must be a valid semantic version\n* spec.kubeadmConfigSpec.clusterConfiguration.certBundles: Forbidden: cannot be modified\n* spec.kubeadmConfigSpec.clusterConfiguration.certificatesDir: Forbidden: cannot be modified\n* spec.kubeadmConfigSpec.initConfiguration: Forbidden: cannot be modified\n* spec.kubeadmConfigSpec.joinConfiguration: Forbidden: cannot be modified\n* spec.version: Internal error: failed to parse updated kubeadmcontrolplane version: v: failed to parse major.minor.patch from \"v\"\n","retries":1}
`



Commands run : 
1. Created cluster using: eksctl anywhere create cluster -f ./cluster.yaml -z ./hardware.csv --force-cleanup --no-timeouts

[cluster.yaml.txt](https://github.com/aws/eks-anywhere/files/11928524/cluster.yaml.txt)
[hardware.csv.txt](https://github.com/aws/eks-anywhere/files/11928527/hardware.csv.txt)


2. Scale cluster using: eksctl anywhere upgrade cluster -f ./cluster-scale.yaml -z ./hardware-scale.csv --force-cleanup --no-timeouts

[cluster-scale.yaml.txt](https://github.com/aws/eks-anywhere/files/11928523/cluster-scale.yaml.txt)
[hardware-scale.csv.txt](https://github.com/aws/eks-anywhere/files/11928526/hardware-scale.csv.txt)

Had to Ctl-C after noticing no tinkerbell workflows  were being created for new node and errors in cli logs (mentioned above)

Output:
`# eksctl anywhere upgrade cluster -f ./cluster-scale.yaml -z ./hardware-scale.csv --force-cleanup --no-timeouts
Warning: The recommended number of control plane nodes is 3 or 5
Warning: The recommended number of control plane nodes is 3 or 5
Warning: The recommended number of control plane nodes is 3 or 5
Performing setup and validations
✅ Tinkerbell provider validation
✅ Validate OS is compatible with registry mirror configuration
✅ Validate certificate for registry mirror
✅ Control plane ready
✅ Worker nodes ready
✅ Nodes ready
✅ Cluster CRDs ready
✅ Cluster object present on workload cluster
✅ Upgrade cluster kubernetes version increment
✅ Validate authentication for git provider
✅ Validate immutable fields
✅ Validate pod disruption budgets
Ensuring etcd CAPI providers exist on management cluster before upgrade
Pausing EKS-A cluster controller reconcile
Pausing GitOps cluster resources reconcile
Upgrading core components
Creating bootstrap cluster
Provider specific pre-capi-install-setup on bootstrap cluster
Provider specific post-setup
Installing cluster-api providers on bootstrap cluster
Backing up workload cluster's management resources before moving to bootstrap cluster
Moving management cluster from workload to bootstrap cluster
Upgrading workload cluster
^CWarning: Terminating this operation may leave the cluster in an irrecoverable state
`

**Comment by robbierafay** [comment]
This issue looks similar to https://github.com/aws/eks-anywhere/issues/6082

```
Currently the Tinkerbell provider splits the cluster spec by "---" to iterate over its items in the omitTinkerbellMachineTemplate() function. When a Registry Mirror CA is defined, this fails, because the CA definition also contains this pattern:
```

**Comment by chrisdoherty4** [comment]
@HiteshRepo Thank you for raising the PR with a test. The patch from #6082 will be available in v0.16.2.

**Comment by HiteshRepo** [comment]
Fix available in [v0.16.2](https://github.com/aws/eks-anywhere/releases/tag/v0.16.2).

**Comment by HiteshRepo** [comment]
@chrisdoherty4 could you please close this?
